File: CommonCode/AudioSource.m
//
//  AudioSource.m
//  Spectrogeddon
//
//  Created by Tom York on 14/04/2014.
//  
//

#import "AudioSource.h"
#import "TimeSequence.h"
#import "SampleBuffer.h"
#import <AVFoundation/AVFoundation.h>

#if TARGET_OS_IPHONE
static const NSUInteger BufferSize = 1024;
static const NSUInteger ReadInterval = 1;
#else
static const NSUInteger BufferSize = 4096;
static const NSUInteger ReadInterval = 8;
#endif

@interface AudioSource ()  <AVCaptureAudioDataOutputSampleBufferDelegate>
@property (nonatomic,strong) dispatch_queue_t sampleQueue;
@property (nonatomic,strong) AVCaptureSession* captureSession;
@property (nonatomic,strong) NSMutableArray* channelBuffers;
@property (nonatomic) NSUInteger channels;
@end

@implementation AudioSource

+ (NSDictionary*)availableAudioSources
{
    NSArray* devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeAudio];
    NSArray* localizedNames = [devices valueForKeyPath:@"localizedName"];
    NSArray* uniqueIDs = [devices valueForKeyPath:@"uniqueID"];
    if(!uniqueIDs || !localizedNames || uniqueIDs.count != localizedNames.count)
    {
        return nil;
    }
    return [NSDictionary dictionaryWithObjects:uniqueIDs forKeys:localizedNames];
}

+ (void)requestPermissionToUseAudio:(void(^)(BOOL isAllowed))permissionBlock
{
    NSParameterAssert(permissionBlock);
    
#if TARGET_OS_IPHONE
    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
        permissionBlock(granted);
    }];
#else
    permissionBlock(YES);
#endif
}

#pragma mark - Lifecycle - 

- (id)initWithNotificationQueue:(dispatch_queue_t)queue block:(AudioSourceNotificationBlock)block
{
    NSParameterAssert(block);
    if((self = [super init]))
    {
        _notificationQueue = queue ?: dispatch_get_main_queue();
        _notificationBlock = [block copy];
        
        _sampleQueue = dispatch_queue_create("com.spectrogeddon.audio.samples", DISPATCH_QUEUE_SERIAL);
        _channelBuffers = [[NSMutableArray alloc] init];
        NSError* error = nil;
#if TARGET_OS_IPHONE
        if(![self prepareAudioSessionWithError:&error])
        {
            DLOG(@"Failed to prepare audio session: %@", error);
            self = nil;
            return nil;
        }
#endif
        if(![self prepareCaptureSessionWithError:&error])
        {
            DLOG(@"Failed to prepare capture session: %@", error);
            self = nil;
            return nil;
        }
    }
    return self;
}

- (void)dealloc
{
#if TARGET_OS_IPHONE
    NSError* error = nil;
    if([self closeAudioSessionWithError:&error]) {
        DLOG(@"Failed to close audio session: %@", error);
    }
#endif
}

#if TARGET_OS_IPHONE
- (BOOL)closeAudioSessionWithError:(NSError**)error
{
    AVAudioSession* session = [AVAudioSession sharedInstance];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:nil object:session];
    return [session setActive:NO withOptions:0 error:error];
}
#endif

#if TARGET_OS_IPHONE
- (BOOL)prepareAudioSessionWithError:(NSError**)error
{
    AVAudioSession* session = [AVAudioSession sharedInstance];
    if(![session setCategory:AVAudioSessionCategoryPlayAndRecord withOptions:AVAudioSessionCategoryOptionMixWithOthers error:error])
    {
        return NO;
    }
    
    if(![session setMode:AVAudioSessionModeMeasurement error:error])
    {
        return NO;
    }
    
    const BOOL success = [session setActive:YES withOptions:0 error:error];
    if(success) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didReceiveAudioInterruption:) name:AVAudioSessionInterruptionNotification object:session];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(audioServicesDidReset:) name:AVAudioSessionMediaServicesWereResetNotification object:session];
    }
    return success;
}
#endif

#if TARGET_OS_IPHONE
- (void)didReceiveAudioInterruption:(NSNotification*)note
{
    DLOG(@"%@", note);
}
#endif

#if TARGET_OS_IPHONE
- (void)audioServicesDidReset:(NSNotification*)note
{
    DLOG(@"%@", note);
}
#endif

- (BOOL)prepareCaptureSessionWithError:(NSError**)error
{
    if(!self.captureSession)
    {
        self.captureSession = [[AVCaptureSession alloc] init];
        self.captureSession.automaticallyConfiguresApplicationAudioSession = NO;
    }

    [self.captureSession beginConfiguration];
    NSArray* existingInputs = self.captureSession.inputs;
    for(AVCaptureInput* oneInput in existingInputs)
    {
        [self.captureSession removeInput:oneInput];
    }
    
    AVCaptureDevice* preferredDevice = [AVCaptureDevice deviceWithUniqueID:self.preferredAudioSourceID];
    if(!preferredDevice)
    {
        preferredDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeAudio];
    }
    
    AVCaptureDeviceInput* micInput = [AVCaptureDeviceInput deviceInputWithDevice:preferredDevice error:error];
    if(!micInput || ![self.captureSession canAddInput:micInput])
    {
        return NO;
    }
    [self.captureSession addInput:micInput];
    
    if(!self.captureSession.outputs.count)
    {
        AVCaptureAudioDataOutput* dataOutput = [[AVCaptureAudioDataOutput alloc] init];
        [dataOutput setSampleBufferDelegate:self queue:self.sampleQueue];
        if(![self.captureSession canAddOutput:dataOutput])
        {
            return NO;
        }
        [self.captureSession addOutput:dataOutput];
    }
    [self.captureSession commitConfiguration];
    return YES;
}

- (void)setPreferredAudioSourceID:(NSString *)preferredAudioSourceID
{
    if(_preferredAudioSourceID != preferredAudioSourceID)
    {
        _preferredAudioSourceID = [preferredAudioSourceID copy];
        [self prepareCaptureSessionWithError:nil];
    }
}

#pragma mark - Capturing

- (void)startCapturing
{
    if(!self.captureSession.isRunning) {
        [self.captureSession startRunning];        
    }
}

- (void)stopCapturing
{
    [self.captureSession stopRunning];
}

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection
{
    const CMItemCount numSamples = CMSampleBufferGetNumSamples(sampleBuffer);
    if(numSamples <= 0)
    {
        return;
    }
    
    // Configure buffers
    // TODO: very basic format decoding, really the minimum.
    BOOL isNormalizedFloatBuffer = NO;
    CMFormatDescriptionRef formatInfo = CMSampleBufferGetFormatDescription(sampleBuffer);
    size_t formatListSize = 0;
    const AudioFormatListItem* bufferFormat = CMAudioFormatDescriptionGetFormatList(formatInfo, &formatListSize);
    if(formatListSize > 0 && bufferFormat)
    {
        isNormalizedFloatBuffer = bufferFormat[0].mASBD.mBytesPerFrame == sizeof(float);
    }

    const NSUInteger channelsInBuffer = bufferFormat[0].mASBD.mChannelsPerFrame;
    if(channelsInBuffer != self.channels)
    {
        for(NSUInteger channelIndex = channelsInBuffer; channelIndex < self.channels; channelIndex++)
        {
            [self.channelBuffers removeObjectAtIndex:0];
        }
        
        for(NSUInteger channelIndex = self.channels; channelIndex < channelsInBuffer; channelIndex++)
        {
            [self.channelBuffers addObject:[[SampleBuffer alloc] initWithBufferSize:BufferSize readInterval:ReadInterval]];
        }
        self.channels = channelsInBuffer;
    }

    // Extract data
	const CMBlockBufferRef audioBlockBuffer = CMSampleBufferGetDataBuffer(sampleBuffer);
    const NSTimeInterval duration = CMTimeGetSeconds(CMSampleBufferGetDuration(sampleBuffer));
    const NSTimeInterval timeStamp = CMTimeGetSeconds(CMSampleBufferGetOutputPresentationTimeStamp(sampleBuffer));

    NSUInteger activeChannel = 0;
    size_t offset = 0;
	size_t totalLength = 0;
    size_t lengthAtOffset = 0;
	char* rawSamples = NULL;
    do {
        CMBlockBufferGetDataPointer(audioBlockBuffer, offset, &lengthAtOffset, &totalLength, &rawSamples);
        if(isNormalizedFloatBuffer)
        {
            [self.channelBuffers[activeChannel] writeNormalizedFloatSamples:(float*)rawSamples count:numSamples timeStamp:timeStamp duration:duration];
        }
        else
        {
            [self.channelBuffers[activeChannel] writeSInt16Samples:(SInt16*)rawSamples count:numSamples timeStamp:timeStamp duration:duration];
        }
        offset += lengthAtOffset;
        activeChannel++;
        
    } while(offset < totalLength);

    // Dispatch
    for(SampleBuffer* oneBuffer in self.channelBuffers)
    {
        if(!oneBuffer.hasOutput)
        {
            return;
        }
    }

    NSMutableArray* outputs = [[NSMutableArray alloc] init];
    for(SampleBuffer* oneBuffer in self.channelBuffers)
    {
        [outputs addObject:[oneBuffer readOutputSamples]];
    }
    dispatch_async(self.notificationQueue, ^{
        self.notificationBlock(outputs);
    });
}

@end

File: CommonCode/AudioSource.swift
//
//  AudioSource.swift
//  Spectrogeddon
//
//  Created by Tom York on 16/05/2019.
//  Copyright © 2019 Random. All rights reserved.
//

import Foundation
import AVFoundation


private let MaxBufferSize: UInt = 2048
private let ReadInterval: UInt = 1

public class AudioSource: NSObject {
    
    public typealias Name = String
    public typealias Identifier = String
    
    /// Localized name -> audio source ID pairs
    public static var availableAudioSources: [Name: Identifier] {
        let devices = AVCaptureDevice.devices(for: .audio)
        let namesAndIds = devices.map { ($0.localizedName, $0.uniqueID) }
        return [Name: Identifier](uniqueKeysWithValues: namesAndIds)
    }
    
    public static func requestMicrophoneAccess(_ completion: @escaping (Bool) -> Void) {
        #if os(iOS)
            AVAudioSession.sharedInstance().requestRecordPermission(completion)
        #else
            completion(true)
        #endif
    }
    
    public private(set) var notificationHandler: ([TimeSequence]) -> Void
    
    public var preferredAudioSourceId: String? {
        didSet {
            try? prepareCaptureSession() // TODO:
        }
    }
    
    public var bufferSizeDivider: UInt {
        didSet {
            isPendingBufferSizeChange = true
        }
    }

    private var bufferSize: UInt {
        return MaxBufferSize / bufferSizeDivider;
    }

    private var notificationQueue: DispatchQueue
    private var sampleQueue: DispatchQueue
    private var captureSession: AVCaptureSession
    private var channelBuffers: [SampleBuffer]
    private var isPendingBufferSizeChange: Bool
    
    init(notificationQueue: DispatchQueue = .main, handler: @escaping ([TimeSequence]) -> Void) throws {
        
        self.notificationQueue = notificationQueue
        self.notificationHandler = handler
        
        self.bufferSizeDivider = 1;
        self.notificationQueue = notificationQueue
        self.notificationHandler = handler

        self.sampleQueue = DispatchQueue(label: "audio.samples", qos: .default)
        self.channelBuffers = []
        self.captureSession = AVCaptureSession()
        self.isPendingBufferSizeChange = false
        ///self.preferredAudioSourceId = AudioSource.availableAudioSources.values.first
        
        super.init()
        
        #if os(iOS)
        try prepareAudioSession()
        #endif

        try prepareCaptureSession()
    }
    
    public func startCapturing() {
        captureSession.startRunning()
    }
    
    public func stopCapturing() {
        captureSession.stopRunning()
    }

    #if os(iOS)
    
    private func prepareAudioSession() throws {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playAndRecord, options: [.mixWithOthers])
        try session.setMode(.measurement)
        try session.setActive(true, options: [])
        
        NotificationCenter.default.addObserver(self, selector: #selector(didReceiveAudioInterruption(_:)), name: AVAudioSession.interruptionNotification, object: session)
        NotificationCenter.default.addObserver(self, selector: #selector(audioServicesDidReset(_:)), name: AVAudioSession.mediaServicesWereResetNotification, object: session)
        NotificationCenter.default.addObserver(self, selector: #selector(audioRouteDidChange(_:)), name: AVAudioSession.routeChangeNotification, object: session)
    }
    
    @objc
    private func didReceiveAudioInterruption(_ note: Notification) {
        print("%@", note);
    }

    @objc
    private func audioServicesDidReset(_ note: Notification) {
        print("%@", note);
    }

    @objc
    private func audioRouteDidChange(_ note: Notification) {
        print("%@", note);
    }

    #endif
    
    private func prepareCaptureSession() throws {
        
        guard let sourceId = preferredAudioSourceId else {
            return
        }
        
        #if os(iOS)
            captureSession.automaticallyConfiguresApplicationAudioSession = false
        #endif
        
        captureSession.beginConfiguration()
        
        captureSession.inputs.forEach {
            captureSession.removeInput($0)
        }
        
        guard
            let preferredDevice = AVCaptureDevice(uniqueID: sourceId) ?? AVCaptureDevice.default(for: .audio) else {
            return // TODO: throw
        }
        
        let micInput = try AVCaptureDeviceInput(device: preferredDevice)
        
        guard captureSession.canAddInput(micInput) else {
            return // TODO: throw
        }
        captureSession.addInput(micInput)
        
        if captureSession.outputs.isEmpty {
            let dataOutput = AVCaptureAudioDataOutput()
            dataOutput.setSampleBufferDelegate(self, queue: sampleQueue)
            
            guard captureSession.canAddOutput(dataOutput) else {
                return // TODO: throw
            }
            
            captureSession.addOutput(dataOutput)
        }
        
        captureSession.commitConfiguration()
    }
}

extension AudioSource: AVCaptureAudioDataOutputSampleBufferDelegate {
 
    public func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {

        let numberOfSamples = CMSampleBufferGetNumSamples(sampleBuffer)
        guard numberOfSamples > 0 else {
            // No samples, no processing.
            return
        }
        
        // Configure buffers
        // TODO: very basic format decoding, really the minimum.
        guard let formatInfo = CMSampleBufferGetFormatDescription(sampleBuffer) else {
            return
        }
        var formatListSize = 0
        guard let bufferFormats = CMAudioFormatDescriptionGetFormatList(formatInfo, sizeOut: &formatListSize), formatListSize > 0 else {
            // Invalid format info.
            return
        }

        let firstFormat = bufferFormats[0]
        
        if isPendingBufferSizeChange {
            channelBuffers = []
            isPendingBufferSizeChange = false
        }
    
        let isNormalizedFloat = firstFormat.mASBD.mBytesPerFrame == MemoryLayout<Float>.size
        let channelsInBuffer = Int(firstFormat.mASBD.mChannelsPerFrame)
        if channelsInBuffer != channelBuffers.count {
            let buffers = (0..<channelsInBuffer).map { _ in
                SampleBuffer(capacity: bufferSize, readInterval: ReadInterval)
            }
            channelBuffers = buffers
        }
        
        // Extract data
        guard let audioBlockBuffer = CMSampleBufferGetDataBuffer(sampleBuffer) else {
            return
        }
        let duration = CMTimeGetSeconds(CMSampleBufferGetOutputDuration(sampleBuffer)) // Duration
        let timeStamp = CMTimeGetSeconds(CMSampleBufferGetOutputPresentationTimeStamp(sampleBuffer))
        
        var activeChannel = 0
        var offset = 0
        var totalLength = 0
        var lengthAtOffset = 0

        repeat {
            var rawSamplesOrNil: UnsafeMutablePointer<Int8>? = nil
            CMBlockBufferGetDataPointer(audioBlockBuffer, atOffset: offset, lengthAtOffsetOut: &lengthAtOffset, totalLengthOut: &totalLength, dataPointerOut: &rawSamplesOrNil)

            guard let rawSamples = rawSamplesOrNil else {
                break
            }
            
            if isNormalizedFloat {
                let asFloats = UnsafeRawPointer(rawSamples).bindMemory(to: Float.self, capacity: numberOfSamples)
                channelBuffers[activeChannel].writeSamples(asFloats, count: UInt(numberOfSamples), timeStamp: timeStamp, duration: duration)
            } else {
                let asWords = UnsafeRawPointer(rawSamples).bindMemory(to: Int16.self, capacity: numberOfSamples)
                channelBuffers[activeChannel].writeSamples(asWords, count: UInt(numberOfSamples), timeStamp: timeStamp, duration: duration)
            }
            offset += lengthAtOffset;
            activeChannel += 1

        } while(offset < totalLength);
    
        // Dispatch
        guard channelBuffers.allSatisfy({ $0.hasOutput }) else {
            return
        }

        let outputs: [TimeSequence] = channelBuffers.compactMap {
            $0.outputSamples
        }

        notificationQueue.async {
            self.notificationHandler(outputs)
        }
    }

}
